gdxsv/lbs_discord.go

```diff
@@ -0,0 +1,525 @@
+			Title:     fmt.Sprintf("**同時接続数 %d 人 **", lobbyUserCount+battleUserCount),
+			Color:     52224,
+			Footer:    &discordEmbedFooter{Text: "🕒"},
+			Timestamp: fmt.Sprintf(time.Now().UTC().Format("2006-01-02T15:04:05.000Z")),
+		})
+
+		//
+		//2nd Embed, lobby count
+		//
+
+		//1st Field is always Plaza
+		var lobbyFields []*discordEmbedField
+		if plazaUserCount > 0 {
+			for i, v := range splitEmbedFieldValues(plazaUsers) {
+				name := "⠀"
+				if i == 0 {
+					name = fmt.Sprintf("**Plaza － %d 人**", plazaUserCount)
+				}
+
+				lobbyFields = append(lobbyFields, &discordEmbedField{
+					Name:  name,
+					Value: v,
+				})
+				accumulatedEmbedStringLength += len(name) + len(v)
+			}
+		}
+
+		//Following Fields would be all lobbies
+		//Use sortedKeys to fix the ordering
+		for _, i := range sortedKeys(lobby) {
+			l := lobby[i]
+
+			value := addBlankIfRequired(l.RenpoUsersString+l.ZeonUsersString) + addBlankIfRequired(l.RenpoRoomUsersString+l.ZeonRoomUsersString) + addBlankIfRequired(l.NoForceUsersString)
+
+			for i, v := range splitEmbedFieldValues(value) {
+				name := "⠀"
+				if i == 0 {
+					name = fmt.Sprintf("**%s － %d 人\n%s %s**", l.Name, l.Count, l.RegionName, l.Comment)
+				}
+
+				lobbyFields = append(lobbyFields, &discordEmbedField{
+					Name:  name,
+					Value: v,
+				})
+				accumulatedEmbedStringLength += len(name) + len(v)
+			}
+
+		}
+		for i, fields := range splitEmbedFields(lobbyFields) {
+			description := ""
+			if i == 0 {
+				description = fmt.Sprintf("🌐 **待機中 %d 人**", lobbyUserCount)
+			}
+			payload.Embed = append(payload.Embed, &discordEmbed{
+				Description: description,
+				Color:       24041,
+				Fields:      fields,
+			})
+			accumulatedEmbedStringLength += len(description)
+		}
+
+		//
+		//3rd Embed, battle count
+		//
+		var battleFields []*discordEmbedField
+		for _, b := range battle {
+
+			value := addBlankIfRequired(b.RenpoUsersString + b.ZeonUsersString)
+
+			for i, v := range splitEmbedFieldValues(value) {
+				name := "⠀"
+				if i == 0 {
+					name = b.RegionName
+				}
+
+				battleFields = append(battleFields, &discordEmbedField{
+					Name:  name,
+					Value: v,
+				})
+				accumulatedEmbedStringLength += len(name) + len(v)
+			}
+		}
+		for i, fields := range splitEmbedFields(battleFields) {
+			description := ""
+			if i == 0 {
+				description = fmt.Sprintf("💥 **戦闘中 %d 人**", battleUserCount)
+			}
+			payload.Embed = append(payload.Embed, &discordEmbed{
+				Description: description,
+				Color:       13179394,
+				Fields:      fields,
+			})
+			accumulatedEmbedStringLength += len(description)
+		}
+
+		//
+		// Create the json
+		//
+		buffer := &bytes.Buffer{}
+		encoder := json.NewEncoder(buffer)
+		encoder.SetEscapeHTML(false)
+		err := encoder.Encode(payload)
+
+		if err != nil {
+			logger.Error("Failed to create Discord JSON", zap.Error(err))
+			return
+		}
+
+		//embed structure must not exceed 6000 characters
+		if accumulatedEmbedStringLength > 6000 {
```

Comment:
If displayed 6000 characters, it will not fit on one screen.
Simply, if more than a certain number of people connected (ex:100 players) to the server, let's omit the details.

Reply:
So I guess the number of people to be displayed may not be relevant to this 6000 char limit already, since it can display 300 people with the UserID and fancy gundam/zaku emoji

I don't think we will ever reach this limit (although I hope), I would be more happy to scroll and see the full list if this happens

Reply:
The current `reduceJSONStringSize()` approach is so heavy that I don't want to introduce it.

Reply:
With regex + replacer

```
**同時接続数 128 人 **  //just over 6000 char
Discord payload size	{"accumulatedEmbedStringLength": 6024} 
Reduce JSON Size!	{"original size": 6435, "elapsed": "130.821µs"}
```

```
**同時接続数 1100 人 **  //maximum supported players
Discord payload size	{"accumulatedEmbedStringLength": 50871}  
Reduce JSON Size!	{"original size": 53293, "elapsed": "1.099318ms"}
```

With replacer only:
```
**同時接続数 128 人 **
Discord payload size	{"accumulatedEmbedStringLength": 50871}
Reduce JSON Size!	{"original size": 53293, "elapsed": "38.312µs"}
```

```
**同時接続数 420 人 **  //maximum supported players with UserID
Discord payload size	{"accumulatedEmbedStringLength": 19495}
Reduce JSON Size!	{"original size": 20497, "elapsed": "73.098µs"}
```

Without removing the `UserID`, the calculation is instant now
let's support 400 concurrent users with `UserID`

Reply:
The `replacer` object can be reused.

Reply:
Done
Commit: [alloc emoji replacer once only](https://github.com/inada-s/gdxsv/pull/66/commits/52924c3e42a58d769664b8305ecb609ab9f4388f)